codec-http/src/main/java/io/netty/handler/codec/http/HttpServerExpectContinueHandler.java

```diff
@@ -0,0 +1,93 @@
+import static io.netty.handler.codec.http.HttpResponseStatus.CONTINUE;
+import static io.netty.handler.codec.http.HttpVersion.HTTP_1_1;
+
+/**
+ * HttpServerExpectContinueHandler sends a <a href="https://www.w3.org/Protocols/rfc2616/rfc2616-sec8.html#sec8.2.3">
+ * 100 CONTINUE</a> {@link HttpResponse} to {@link HttpRequest}s which contain a 'expect: 100-continue' header. It
+ * should only be used for applications which do <b>not</b> install the {@link HttpObjectAggregator}.
+ * <p>
+ * By default it accepts all expectations.
+ * <p>
+ * Since {@link HttpServerExpectContinueHandler} expects {@link HttpRequest}s it should be added after {@link
+ * HttpServerCodec} but before any other handlers that might send a {@link HttpResponse}. <blockquote>
+ * <pre>
+ *  {@link io.netty.channel.ChannelPipeline} p = ...;
+ *  ...
+ *  p.addLast("serverCodec", new {@link HttpServerCodec}());
+ *  p.addLast("respondExpectContinue", <b>new {@link HttpServerExpectContinueHandler}()</b>);
+ *  ...
+ *  p.addLast("handler", new HttpRequestHandler());
+ *  </pre>
+ * </blockquote>
+ */
+public class HttpServerExpectContinueHandler extends ChannelInboundHandlerAdapter {
+
+    private static final FullHttpResponse EXPECTATION_FAILED = new DefaultFullHttpResponse(
+            HTTP_1_1, HttpResponseStatus.EXPECTATION_FAILED, Unpooled.EMPTY_BUFFER);
+
+    static {
+        EXPECTATION_FAILED.headers().set(CONTENT_LENGTH, 0);
+    }
+
+    /**
+     * Produces a {@link HttpResponse} for {@link HttpRequest}s which define an expectation. Returns {@code null} if the
+     * request should be rejected. See {@link #rejectResponse(HttpRequest)}.
+     */
+    protected HttpResponse acceptMessage(@SuppressWarnings("unused") HttpRequest request) {
+        return new DefaultFullHttpResponse(HTTP_1_1, CONTINUE);
+    }
+
+    /**
+     * Returns the appropriate 4XX {@link HttpResponse} for the given {@link HttpRequest}.
+     */
+    protected HttpResponse rejectResponse(@SuppressWarnings("unused") HttpRequest request) {
+        return EXPECTATION_FAILED.retainedDuplicate();
+    }
+
+    @Override
+    public void channelRead(ChannelHandlerContext ctx, Object msg) throws Exception {
+        if (msg instanceof HttpRequest) {
+            HttpRequest req = (HttpRequest) msg;
+
+            if (HttpUtil.is100ContinueExpected(req)) {
+                HttpResponse accept = acceptMessage(req);
+                if (accept != null) {
+                    ctx.writeAndFlush(accept).addListener(ChannelFutureListener.CLOSE_ON_FAILURE);
+                    req.headers().remove(HttpHeaderNames.EXPECT);
+                    super.channelRead(ctx, msg);
+                } else {
+                    // the expection failed so we refuse the request.
+                    HttpResponse rejection = rejectResponse(req);
+                    ReferenceCountUtil.release(msg);
+                    ctx.writeAndFlush(rejection).addListener(ChannelFutureListener.CLOSE_ON_FAILURE);
```

Comment:
How about early-returning here and move the call `super.channelRead(ctx, msg)` to the bottom of the method? i.e.

```java
if (msg instanceof HttpRequest) {
    if (HttpUtil.is100ContinueExpected(req)) {
        if (accept == null) {
            ctx.writeAndFlush(rejection)....;
            return;
        }
    }
}

super.channelRead(ctx, msg);
```